[elhy1999] iP

elhy1999 opened the pull request:
# TaskMate
> My basic idea is that play and productivity *aren't opposed* to one another -
> Ali Abdaal

A ~~normal~~ **SUPER COOL** task-management tool for your everyday needs!

## Features
* Add different types of tasks: todo, deadline, and event
* Mark your tasks as completed or incomplete
* Find tasks by name

## Steps to Download
1. Go to iP repository and click on "[Releases](https://github.com/elhy1999/ip/releases)"
2. Choose the version you want and download the `.jar` file
3. Run: `java -jar ip.jar`

## Simulation of TaskMate:
1. [ ] todo: Read Book
2. [ ] deadline: return book (by: 27 Sep 2023)
3. [ ] event: attend music festival 💯 : (from: 6 Sep 2023 to: 7 Sep 2023)

## Prof say must write fenced code block:
```java
public static void main(String[] args) {
    System.out.println("Hello world");
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

elhy1999 committed:
Changed Chatbot name from Duke to TaskMate. Greets user and exits

elhy1999 committed:
Chatbot echoes user input until bye

elhy1999 committed:
Added Task class to store all task-related information

elhy1999 committed:
Chatbot echoes user input and stores an arraylist of tasks set by the user. Able to view the list with the list command

elhy1999 committed:
Included isDone variable in Task class. Not yet able to allow users to mark and unmark tasks

elhy1999 committed:
Chatbot is able to mark and unmark tasks, and prints messages when marking/unmarking

elhy1999 committed:
Created functions to handle commands: list, mark/unmark, add task

elhy1999 committed:
Added Deadline, Event and Todo child classes of Task class

elhy1999 committed:
Modified Task and TaskMate classes to handle different types of tasks.

elhy1999 committed:
Added test case for level-4

elhy1999 committed:
Made minor changes to TaskMate.java. Simple indentation changes

elhy1999 committed:
Created InvalidCommandTypeException class for commands that are not in the chatbot's vocabulary

elhy1999 committed:
Handled invalid command exception. Invalid description not completed

elhy1999 committed:
Handled invalid description exception only for mark/unmark commands, todo command. Not yet completed for other commands

elhy1999 committed:
Edited EXPECTED.txt according to test case

elhy1999 committed:
Edited code to use enums instead of checking for String equality

elhy1999 committed:
Added delete functionality with checks for potential wrong inputs in place

elhy1999 committed:
Added another test case with delete operation

elhy1999 committed:
Add functionality to Task class to keep track of the number of incomplete tasks.

elhy1999 committed:
Add functionality to write tasks to disk

elhy1999 committed:
Add functionality to load existing tasks saved in root/data/ directory.

elhy1999 committed:
Merge branch 'branch-Level-7'

elhy1999 committed:
Edit saved_tasks.txt to contain valid dates for by, from and to clauses

elhy1999 committed:
Modify Deadline and Event classes to parse and store dates

elhy1999 committed:
Merge branch 'branch-Level-8'

elhy1999 committed:
Fix bug for reading saved_tasks.txt

Had an error due to date formatting. Bug fixed and now, displays the date in format requested, but stored in disk using simpler format

elhy1999 committed:
Abstracted Ui and TaskList away from TaskMate class. Fixing bug which does not save whether a task is marked or not

elhy1999 committed:
Fix bug to correctly save and load whether a task is marked or not

elhy1999 committed:
Edit Storage to be able to write to disk

elhy1999 committed:
Abstracted Storage away from TaskMate class. Working on Parser now.

elhy1999 committed:
Change UI and TaskMate code to use UI object to read user input.

elhy1999 committed:
Abstract away commands into different Command classes and its children

elhy1999 committed:
Abstract away Parser. Add new exception classes to illicit wider range of behaviours in chatbot

elhy1999 committed:
Minor change to TaskMate class. Remove horizontal line attribute as UI tasks are offloaded to Ui class

elhy1999 committed:
Merge branch 'branch-A-MoreOOP'

elhy1999 committed:
Package classes into different packages

elhy1999 committed:
Merge branch 'branch-A-Packages'

elhy1999 committed:
Merge branch 'add-gradle-support'

elhy1999 committed:
Automate project build with Gradle

elhy1999 committed:
Add unit tests using JUnit. Bug fixes to Deadline, Event and TaskList classes

elhy1999 committed:
Merge branch 'branch-A-JUnit'

elhy1999 committed:
Create JavaDoc for all Command classes

elhy1999 committed:
Create JavaDoc for all Exception classes

elhy1999 committed:
Create JavaDoc for all Task classes

elhy1999 committed:
Edit all Command classes to Java Coding Standard

elhy1999 committed:
Edit all classes in tools package to Java Coding Standard

elhy1999 committed:
Edit TaskMate.java to Java Coding Standard

elhy1999 committed:
Create FindCommand class

elhy1999 committed:
Modify Ui class by adding function to print result for find command

elhy1999 committed:
Change getAllTasks() function's access modifier in TaskList class from default to public

elhy1999 committed:
Edit Parser class to parse find commands

elhy1999 committed:
Merge branch 'branch-A-JavaDoc'

elhy1999 committed:
Fix merge conflict with branch-A-CodingStandard

elhy1999 committed:
Fix merge conflict with branch-Level-9

leezhengjing commented on src/main/java/taskmate/tools/Ui.java:
```diff
@@ -0,0 +1,160 @@
+
+import java.util.ArrayList;
+import java.util.Scanner;
+
```
Maybe there should be a class description?

elhy1999 replied:
Good idea! Will get to that soon thanks for the suggestion 😄

leezhengjing commented on src/main/java/taskmate/tools/Ui.java:
```diff
@@ -0,0 +1,160 @@
+
+    public void printEmptyToExceptionResponse() {
+        printMessage("To clause cannot be empty!");
+    }
+
+    public void printEmptyFromExceptionResponse() {
+        printMessage("From clause cannot be empty!");
+    }
+
+    public void printInvalidByExceptionResponse() {
+        printMessage("By clause must be in the following format: YYYY-MM-DD");
+    }
+
+    public void printInvalidToExceptionResponse() {
+        printMessage("To clause must be in the following format: YYYY-MM-DD");
+    }
+
+    public void printInvalidFromExceptionResponse() {
+        printMessage("From clause must be in the following format: YYYY-MM-DD");
+    }
+
+    public void printEmptyTodoDescriptionResponse() {
+        printMessage("☹ OOPS!!! The description of a todo cannot be empty.");
+    }
+
+    public void printInvalidDeleteResponse(int numTotalTasks) {
+        printMessage("☹ OOPS!!! The description of a delete must be between 1 and " + numTotalTasks + ".");
+    }
+
+    public void printNotAnIntegerExceptionResponse() {
+        printMessage("Please enter a valid integer (E.g. mark 1, unmark 8, delete 3)");
+    }
+
+    public void printFileNotFoundResponse(String filePath) {
+        printMessage("No previous datafile found in " + filePath + ". Creating new task list for you!");
+    }
+
+    public void printTaskNotFoundExceptionResponse() {
+        printMessage("☹ OOPS!!! No such task exists in your task list");
+    }
```
I like how your exception handling is very thorough!

leezhengjing commented on src/main/java/taskmate/tools/Parser.java:
```diff
@@ -0,0 +1,224 @@
+            indexToMark -= 1;
+            return new MarkCommand(indexToMark);
+        // to-do description
+        } else if (commandType.equals(TaskMate.COMMAND_TYPES.todo.toString())) {
+            checkValidTodoCommand(userInput);
+            return new TodoCommand(userInput.substring(TaskMate.COMMAND_TYPES.todo.toString().length()+1)); // +1 because we do not want the task name to start from the space character
+        // deadline description /by date
+        } else if (commandType.equals(TaskMate.COMMAND_TYPES.deadline.toString())) {
+            checkValidDeadlineCommand(userInput);
+            userInput = userInput.substring(TaskMate.COMMAND_TYPES.deadline.toString().length()+1); // +1 because we do not want the task name to start from the space character
+            String[] splitUserInput = userInput.split(" /");
+            return new DeadlineCommand(
+                    splitUserInput[0],
+                    splitUserInput[1].replace("by ", "")
+            );
+        // event description /from date /to date
+        } else if (commandType.equals(TaskMate.COMMAND_TYPES.event.toString())) {
+            checkValidEventCommand(userInput);
+            userInput = userInput.substring(TaskMate.COMMAND_TYPES.event.toString().length()+1); // +1 because we do not want the task name to start from the space character
+            String[] splitUserInput = userInput.split(" /");
+            return new EventCommand(
+                    splitUserInput[0],
+                    splitUserInput[1].replace("from ", ""),
+                    splitUserInput[2].replace("to ", "")
+            );
+        // delete i
+        } else if (commandType.equals(TaskMate.COMMAND_TYPES.delete.toString())) {
+            checkValidDeleteCommand(userInput);
+            int indexToDelete = Integer.parseInt(userInput.substring(TaskMate.COMMAND_TYPES.delete.toString().length()).trim());
+            indexToDelete -= 1; // subtract 1 as the arraylist is zero-indexed
+            return new DeleteCommand(indexToDelete);
+        } else if (commandType.equals(TaskMate.COMMAND_TYPES.find.toString())) {
+            checkValidFindCommand(userInput);
+            String query = userInput.substring(TaskMate.COMMAND_TYPES.find.toString().length()).trim();
+            return new FindCommand(query);
+        // Invalid input
+        } else {
+            throw new InvalidCommandTypeException();
+        }
+    }
```
I like how your code is very well extracted to the different classes!

leezhengjing reviewed (commented):
Overall really high quality individual project, a lot better than what I have done.
Very inspiring to look at your code!

xyT-T commented on src/main/java/taskmate/main/TaskMate.java:
```diff
@@ -0,0 +1,111 @@
+    static String CHATBOT_NAME = "TaskMate";
+    static String DEFAULT_SAVE_TASK_FILE_PATH = "./data/saved_tasks.txt";
+
+    public enum COMMAND_TYPES {
```
I like how you use enum to group all the command words. Maybe you can declare the field as final since you want it to be a constant and name it using capitals?

xyT-T commented on src/main/java/taskmate/main/TaskMate.java:
```diff
@@ -0,0 +1,111 @@
+    private Parser parser;
+
+    TaskMate(String filePath) {
+        this.ui = new Ui(CHATBOT_NAME);
```
I like how you declare the chatbot name as a constant and then pass it into Ui. This allows you to change the name of the chatbot easily

xyT-T commented on src/main/java/taskmate/tools/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        String[] lines = contentsFromDisk.split("\\n");
+        String taskType, name, by, from, to, delimiter, delimiter2;
```
Maybe you can make it clearer by specify the name of the dilimiter instead of calling it delimiter2?

xyT-T commented on src/main/java/taskmate/tools/tasks/Deadline.java:
```diff
@@ -0,0 +1,74 @@
+        return "[D][" + (this.getIsDone() ? 'X' : ' ') + "] " + this.name + " (by: " + this.getByFormatted() + ")";
+    }
+
+    public String formatTaskForSaving() {
```
I like how you naming clearly conveys what the function will do. Good job!😄

elhy1999 committed:
Incorporate JavaFX to TaskMate. Chatbot now echoes user input

elhy1999 committed:
Modify chatbot to have correct response. Does not greet user

elhy1999 committed:
Complete Level-10

elhy1999 committed:
Configure build.gradle to build based on Launcher as main class instead of TaskMate

elhy1999 committed:
Merge branch 'branch-Level-10'

elhy1999 committed:
TaskMate.java: Remove unused function - getGreetResponse

elhy1999 committed:
New Unit Tests: ParserTest.java created for unit testing on Parser class

The getCommand method from the Parser class did not have unit tests
created for it. It had unexpected behaviors when the user inputs their
command with a trailing whitespace.

Let's
* Create unit tests for the getCommand method to ensure future changes
will not cause regressions for this command.
* Create unit tests for the getCommand method to ensure the code is able
to handle trailing whitespaces in the user input.

Allowing for proper handling of trailing whitespaces will allow more
flexibility in the user input, allowing the chatbot to handle more typo
mistakes made by the user.

elhy1999 committed:
Bug fixes on TaskMate.java and Parser.java

Remove whitespaces during processing.

elhy1999 committed:
Merge branch 'branch-add-more-tests'

elhy1999 committed:
Command classes: Write assertions for `execute` method

The `execute` method in some command classes require the command objects
to be correctly initialized. This means that some fields must not be
empty, such as the "by" clause for the `Deadline` objects, and "from"
and "to" clauses for the `Event` objects. The cases where the user
forgets to fill in these clauses should already be handled. These
assertions ensure that future changes to the code base will still allow
them to be handled.

Let's:
* Create `assert` statements for the required fields to be not `null` to
allow the `execute` method to function as expected.

Assertions will ensure that the fields will not be left as empty in the
event that the exceptions fail to be raised.

elhy1999 committed:
Bug fixes: TaskMate, Parser, Ui, ParserTest

elhy1999 committed:
Task classes: Write assertions for constructor methods

The constructor methods in all Task classes require the Task objects
to be correctly initialized. This means that some fields must not be
empty, such as the "by" clause for the `Deadline` objects, and "from"
and "to" clauses for the `Event` objects. The cases where the user
forgets to fill in these clauses should already be handled. These
assertions ensure that future changes to the code base will still allow
them to be handled.

Let's:
* Create `assert` statements for the required fields to be not `null` to
allow proper initialization of Task objects.

Assertions will ensure that the fields will not be left empty in the
event that the exceptions fail to be raised.

elhy1999 committed:
Create assertions for taskmate.tools package

elhy1999 committed:
Remove unused methods in TaskMate.java

elhy1999 committed:
Merge pull request #1 from elhy1999/branch-A-Assertions

Merge branch-A-Assertions into master branch

elhy1999 committed:
Add ./config/ directory for CheckStyle files

elhy1999 committed:
Configure build.gradle to use checkstyle files

elhy1999 committed:
Merge branch 'branch-A-CheckStyle'

elhy1999 committed:
Edit Command classes to adhere to CheckStyle

elhy1999 committed:
Edit TaskMate class to adhere to CheckStyle

elhy1999 committed:
Edit Launcher class to adhere to CheckStyle

elhy1999 committed:
Edit Task classes to adhere to CheckStyle

elhy1999 committed:
Edit Main.java to adhere to CheckStyle

elhy1999 committed:
Edit taskmate.tools package to adhere to CheckStyle

elhy1999 committed:
Edit taskmate.ui package to adhere to CheckStyle

elhy1999 committed:
Edit unit tests to adhere to CheckStyle

elhy1999 committed:
Modify .gitignore to include personal tasks list

elhy1999 committed:
Modify .gitignore to include personal tasks list

elhy1999 committed:
TaskMate.java: Extract exception handling from run method to handleException method

elhy1999 committed:
TaskMate.java: Extract exception handling from getResponse method to handleException method

elhy1999 committed:
TaskList.java: Modify TaskList constructor's if-else to switch-case statements

elhy1999 committed:
FileCorruptedException.java: New exception created to be raised if save file is corrupted

elhy1999 committed:
TaskMate.java: Remove wildcard import

elhy1999 committed:
TaskList.java: Extract task creation in TaskList constructor to three separate methods

elhy1999 committed:
TaskList.java: Change some comments to in-line comments as they are short

elhy1999 committed:
Parser.java: Extract DeleteCommand creation from parse method to abide by SLAP

elhy1999 committed:
Parser.java: Extract away creation of EventCommand object from parse method

elhy1999 committed:
Parser.java: Extract away creation of FindCommand object from parse method

elhy1999 committed:
Parser.java: Change some static methods to private as they are not needed by other classes

elhy1999 committed:
Parser.java: Removed '+ 1' in code as it is hard to understand and replace with  method

elhy1999 committed:
Parser.java: Modify JavaDoc to follow conventions

elhy1999 committed:
Storage.java: Modify getDirectoryPath method accessibility to private

elhy1999 committed:
TaskList.java: Modify JavaDoc to follow conventions

elhy1999 committed:
ExitCommand.java: Extract the formatting of filepath for printing to a separate method for SLAP

elhy1999 committed:
Minor fixes: Name changes

elhy1999 committed:
Merge pull request #3 from elhy1999/branch-A-CodeQuality

Branch A-CodeQuality

elhy1999 committed:
Add update functionality outline, to debug later

elhy1999 committed:
Parser.java: Debug getChangesToUpdate method

elhy1999 committed:
Ui.java: Add print messages to handle responses to update command

elhy1999 committed:
Handle invalid input exceptions for update feature

elhy1999 committed:
Enhance update feature to have more varied response to wrong user input

elhy1999 committed:
Parser.java TaskList.java Ui.java Event.java Task.java: Code change to adhere to checkstyle

elhy1999 committed:
UpdateCommand.java: Modify execute method (bug fixes)

elhy1999 committed:
Ui.java: Minor change to printSuccessfulUpdateResponse function for readability

elhy1999 committed:
Ui.java: Add documentation for help command

elhy1999 committed:
Task.java, Deadline.java, Event.java: Refactor function name isValidDateFormat->checkValidDateFormat

elhy1999 committed:
Parser.java: Extract checking of valid datetime format to checkValidDateFormat method

elhy1999 committed:
Create new exception for update commands without any input clause

elhy1999 committed:
Update feature: Handle edge case exception when user makes an update command without any clause

elhy1999 committed:
Parser.java: Bug fixes to checkValidUpdateCommand method

elhy1999 committed:
Bug fixes to TaskMate.java and Ui.java

elhy1999 committed:
Bug fixes to Parser.java: Check for clauses in checkValidUpdateCommand method

elhy1999 committed:
Parser.java: Fix bug for getChangesToUpdate method

elhy1999 committed:
Merge pull request #4 from elhy1999/branch-C-Update

Branch c update

elhy1999 committed:
DialogBox.fxml: Bug fix to make box height configurable to message length

elhy1999 committed:
Change images from default

elhy1999 committed:
Change font family to Courier New, Font size 11

elhy1999 committed:
Ui.java: Remove emojis as they cannot be displayed

elhy1999 committed:
Add margins to sides of images

elhy1999 committed:
Change background colour to #89d9f0 (Light blue)

elhy1999 committed:
Changed background colour to #a4c6fc (baby blue)

elhy1999 committed:
Resize images

elhy1999 committed:
Main.java: Add stage name 'TaskMate Gui'

elhy1999 committed:
Add Ui.png to doc/ directory

elhy1999 committed:
Merge pull request #5 from elhy1999/branch-add-Ui-image

Add Ui.png to doc/ directory

elhy1999 committed:
README.md: Add incomplete User Guide

elhy1999 committed:
README.md: Add descriptions for more commands

elhy1999 committed:
Merge branch 'branch-add-readme'

elhy1999 committed:
Ui.java: Fix typo errors for help page

elhy1999 committed:
Merge branch 'branch-A-BetterGui'

elhy1999 committed:
Add title to README.md

elhy1999 committed:
Merge branch 'branch-A-BetterGui'

elhy1999 committed:
Add TaskMate GUI image to README.md

elhy1999 committed:
Fix merge conflict when merging master into branch-A-BetterGui

elhy1999 committed:
README.md: Rectify a single typo error

elhy1999 committed:
Merge branch 'branch-add-readme'

elhy1999 committed:
README.md: Add 'Downloading TaskMate' section to guide users on how to download TaskMate

elhy1999 committed:
Merge branch 'branch-add-readme'

elhy1999 committed:
TaskListTest.java: Add more unit tests for TaskList#addTask to cover more possible cases

elhy1999 committed:
TaskListTest.java: Modify test cases to make TaskList#addTask more likely to fail

elhy1999 committed:
Merge branch 'branch-A-JUnit'

elhy1999 committed:
ParserTest.java: Add more unit tests to cover cases where encountering typo errors

elhy1999 committed:
Merge branch 'branch-A-JUnit'

elhy1999 committed:
Task.java and children: Extract Task#update into two separate methods for SLAP

elhy1999 committed:
taskmate.tools.tasks package: Add JavaDocs to public methods

elhy1999 committed:
taskmate.commands package: Change the first word of all JavaDocs to be a present-tense verb

elhy1999 committed:
Edit all JavaDocs to start with present-tense verbs

elhy1999 committed:
Merge branch 'branch-A-CodeQuality'

elhy1999 committed:
TaskList.java: Add back empty constructor method as it was wrongly removed, causing errors

elhy1999 committed:
Merge branch 'branch-A-CodeQuality'

elhy1999 committed:
TaskListTest.java: Modify expected test output Strings

elhy1999 committed:
Merge branch 'branch-java-version-change'
